rfcs/0000-type-stability.md

```diff
@@ -0,0 +1,499 @@
+
+### Use an alternative to `downlevel-dts`
+
+As noted above in [<b>“Downleveling” types</b>](#downleveling-types), `downlevel-dts` currently only generates types for TypeScript 3.4, with no ability to generate targeted types. As such, there are two other approaches we might take to 
+
+1.  **Build our own tool.** While this is probably doable—`downlevel-dts` is not especially large or complicated—it is additional maintenance burden, is not aligned with the TypeScript team’s own efforts. It is, as such, not *preferable*. However, if the TypeScript team is unwilling to maintain a more granular tool, this may be appropriate, and it would likely see wide uptake across the broader TypeScript ecosystem, as its utility is obvious.
+
+2.  **Manually generate downleveled types.** This is possible, but very difficult: it requires addon authors to deeply understand the semantics of changes between TypeScript versions, rather than simply allowing a tool to manage it automatically. In practice, we expect that addon authors would choose *not* to downlevel their types because of the complexity involved, and instead either to just wait to adopt *new* TypeScript versions for a longer period of time or to drop support for *older* versions much more rapidly than this RFC proposes, with all the attendant downsides of either.
+
+## Unresolved questions
+
+-   Is the recommended version support policy appropriate? There are other options available, like Typed Ember’s current commitment to support the latest two (<i>N−1</i>) versions in the types. (In practice, we did not bump most of the Ember types’ minimum version from TypeScript 2.8 until the release of TypeScript 3.9, at which time we bumped minimum supported TypeScript version to 3.7.)
+
+-   How should we make downleveling and CI runs against supported TypeScript versions interact? Should we tell users to generate test-specific `tsconfig.json` files to use with downleveled types in their tests? Can this be automated in some way? Does use of `typesVersions` plus the CI runs handle it?
+
+-   What is the right strategy and naming for validating types: "lint," "test," or a new-to-the-ecosystem concept like "check"? (See [this discussion](https://github.com/ember-modifier/ember-modifier/pull/41#discussion_r435628423) for background.)
+
+-   How should transitive dependencies be expected to be handled? Should addon authors be expected to absorb any upstream differences in SemVer handling?
+
+-   Given the prevalence of QUnit in the Ember community, the Jest/Mocha-like `expectType` assertions may seem out of place in some addons. Should we write an `assert-type` library akin to `expect-type`, with assertions like `assertType<SomeType>().equal<{ count: number }>()`?
+
+-   Is a JS file in a repository/package the appropriate home for supported TypeScript versions? Should it also be publicized
+
+-   Should we design a “badge” for addon authors to specify their supported versions of TypeScript?
```

Comment:
We can probably come up with a shields.io URL template to recommend, e.g. ![supported TS versions: 3.6, 3.7 and 3.9](https://img.shields.io/badge/TypeScript-3.6%20%7C%203.7%20%7C%203.9-blue)

```md
![supported TS versions: 3.6, 3.7 and 3.9](https://img.shields.io/badge/TypeScript-3.6%20%7C%203.7%20%7C%203.9-blue)
```

Reply:
❤️ love it!

Reply:
Resolved in e5b4311.